### DISCUSSION TITLE
send() return value of 0

### DISCUSSION THREAD

**Discussion by KrishnaPG** [uNetworking/uWebSockets.js#836] [Q&A] [answered]
The documentation of [send](https://unetworking.github.io/uWebSockets.js/generated/interfaces/WebSocket.html#send) method indicates:
> Returns 1 for success, 2 for dropped due to backpressure limit, and 0 for built up backpressure that will drain over time. 

When the return value is 0: 
   - does it mean the message has been added to the socket buffer queue (and will be sent soon), or should we call `send` again with the same message after the drain event?

I would like to retry the failed messages. Is this below approach correct? I am maintaining a `retryQ[]` to keep track of failed messages so that I can retry them after the drain event. It is not clear how to handle the return value 0.

```js
const retryQ= [];

const sendNow = (ws, msg) => {
  const status = ws.send(msg, true, true);
  switch (status) {
    case 1:  // sent successfully, nothing to do
      return status;
    case 2: { // dropped due to backpressure, lets try again later after drain even
      retryQ.push(msg);
      return status;
    }
    case 0: { // built-up backpressure: what should be done????
      retryQ.push(msg);  //<<-- should we retry the message again or has it already been queued at socket?
      return status;
    }
  }
};
```

**Comment by e3dio** [comment, marked as answer]
```javascript
 case 0: { // built-up backpressure: what should be done????
```
It is already in backpressure queue ready to send out, you don't need to do anything there. Only need to handle the dropped messages

**Reply by KrishnaPG** [reply]
Thank you @e3dio

**Reply by KrishnaPG** [reply]
One follow-up question on this `case 0 (message in the backpressure queue) `:

- how does the socket process the messages that are put in the backpressure queue, since Node.js is single-threaded? Are there threads in the C++ side that take care of these backpressure queued ones, or should something be done from the Node.js side to let the socket process them?

**Reply by uNetworkingAB** [reply]
it's the same as any other nodejs project. you do nothing special
